### ISSUE TITLE
Re-work test verification and add pre-test option in rt.sh for developers to use in PR process.

### ISSUE THREAD

**Issue by BrianCurtis-NOAA** [ufs-community/ufs-weather-model#2058]
## Description
There are a lot of pieces to putting a pull request together and the information that code management needs isn't always readily available and can take developers a long time to gather. The logs are overly verbose and don't provide developers a quick way to find what went wrong with their code changes.

## Solution
The logs should be more succinct. Having information on all file comparisons kept in the history of the repository seems overkill. Simplify the logs to provide information that all tests are pass/fail and provide a log for failed tests for developers to easily find out what went wrong.

## Ideas

- Creates a synopsis that clarifies: 
  - number of tests ran to completion (double checking all tests are run)
    - Listing all failed comparisons.
  - number of compilations ran to completion (double checking all compiles are run).
    - Listing all failed compiles.
- Write synopsis to RegressionTests_<machine>.log
- fail_test uses number while fail_compile uses compile name, changing fail_test to use test name.

## Extras
- The resulting PR will also need to adjust the PR template.

**Comment by BrianCurtis-NOAA** [comment]
As I start working on this, adding tags for @junwang-noaa @DusanJovic-NOAA @DeniseWorthen. Is there anything else you think this should do, or not do, based on what I have in the issue?

**Comment by DeniseWorthen** [comment]
Do you envision that the pre-test log is short (just summarizing and flagging failed tests etc) or would it be like the existing log with a summary attached?

**Comment by BrianCurtis-NOAA** [comment]
> Do you envision that the pre-test log is short (just summarizing and flagging failed tests etc) or would it be like the existing log with a summary attached?

My goal is to limit it to the information we need, and skip the individual file comparisons. If it's not easily fit into the code, it may be larger.

**Comment by DeniseWorthen** [comment]
And is there anyway to have Github understand the pre-test log and apply some sort of "green light" on a PR?

**Comment by BrianCurtis-NOAA** [comment]
> And is there anyway to have Github understand the pre-test log and apply some sort of "green light" on a PR?

I think this would be implemented through a script that the github action uses. We would have to write it, though.

**Comment by BrianCurtis-NOAA** [comment]
I've added code to the script that creates a symbolic link called run_dir that links to the current rt_###### for your running job and put it into the tests dir. It will update that for each new job.

**Comment by BrianCurtis-NOAA** [comment]
This should close #1821

**Comment by BrianCurtis-NOAA** [comment]
```
briancurtis@derecho4:/glade/work/briancurtis/git/BrianCurtis-NOAA/ufs-weather-model/rt_pretest/tests> cat logs/RegressionTests_pretest.log 
Thu 21 Dec 2023 07:21:43 PM MST
Start Regression test

Testing UFSWM Hash: 4d1c08c60c5026945bd4013384e55cb7f8999087
Testing With Submodule Hashes:
 37cbb7d6840ae7515a9a8f0dfd4d89461b3396d1 AQM (v0.2.0-37-g37cbb7d)
 2aa6bfbb62ebeecd7da964b8074f6c3c41c7d1eb CDEPS-interface/CDEPS (cdeps0.4.17-38-g2aa6bfb)
 50aa2c97882fbc9d4918813a22169fe97b424564 CICE-interface/CICE (CICE6.0.0-444-g50aa2c9)
 e5d08d4233b0c783a4840dcbc3252a170e3c3bb1 CMEPS-interface/CMEPS (cmeps_v0.4.1-2299-ge5d08d4)
 cabd7753ae17f7bfcc6dad56daf10868aa51c3f4 CMakeModules (v1.0.0-28-gcabd775)
 f221fc5ce66cee86160efa4bc4deb9c861959e19 FV3 (remotes/origin/production/HREF.v3beta-287-gf221fc5)
 041422934cae1570f2f0e67239d5d89f11c6e1b7 GOCART (sdr_v2.1.2.6-119-g0414229)
 35789c757766e07f688b4c0c7c5229816f224b09 HYCOM-interface/HYCOM (2.3.00-121-g35789c7)
 a36cb73d6924f6cf56a72b5799bef3d75fe4dd61 MOM6-interface/MOM6 (dev/master/repository_split_2014.10.10-9871-ga36cb73d6)
 569e354ababbde7a7cd68647533769a5c966468d NOAHMP-interface/noahmp (v3.7.1-303-g569e354)
 02693d837f2cd99d20ed08515878c2b5e9525e64 WW3 (6.07.1-343-g02693d83)
 37e740526993fd162d092266ced19c625fbc3d8e stochastic_physics (ufs-v2.0.0-193-g37e7405)
Verifying: Compile s2swa_32bit_intel
Verifying: Test cpld_control_p8_mixedmode_intel
Verifying: Compile s2swa_32bit_pdlib_intel
Verifying: Test cpld_control_gfsv17_intel
Verifying: Test cpld_control_gfsv17_iau_intel
Verifying: Test cpld_restart_gfsv17_intel
Verifying: Test cpld_mpi_gfsv17_intel
Verifying: Compile s2swa_32bit_pdlib_debug_intel
Verifying: Test cpld_debug_gfsv17_intel
Verifying: Compile s2swa_intel
Verifying: Test cpld_control_p8_intel
Verifying: Test cpld_restart_p8_intel
Verifying: Test cpld_control_qr_p8_intel
Verifying: Test cpld_restart_qr_p8_intel
Verifying: Test cpld_decomp_p8_intel
Verifying: Test cpld_mpi_p8_intel
Verifying: Test cpld_control_ciceC_p8_intel
Verifying: Test cpld_control_c192_p8_intel
Verifying: Test cpld_restart_c192_p8_intel
Verifying: Test cpld_s2sa_p8_intel
Verifying: Compile s2sw_intel
Verifying: Test cpld_control_noaero_p8_intel
Verifying: Test cpld_control_nowave_noaero_p8_intel
Verifying: Compile s2swa_debug_intel
Verifying: Test cpld_debug_p8_intel
Verifying: Compile s2sw_debug_intel
Verifying: Test cpld_debug_noaero_p8_intel
Verifying: Compile s2s_aoflux_intel
Verifying: Test cpld_control_noaero_p8_agrid_intel
Verifying: Compile s2s_intel
Verifying: Test cpld_control_c48_intel
Verifying: Compile s2sw_pdlib_intel
Verifying: Test cpld_control_pdlib_p8_intel
Verifying: Test cpld_restart_pdlib_p8_intel
Verifying: Test cpld_mpi_pdlib_p8_intel
Verifying: Compile s2sw_pdlib_debug_intel
Verifying: Test cpld_debug_pdlib_p8_intel
Verifying: Compile atm_dyn32_intel
Verifying: Test control_flake_intel
Verifying: Test control_CubedSphereGrid_intel
Verifying: Test control_CubedSphereGrid_parallel_intel
Verifying: Test control_latlon_intel
Verifying: Test control_wrtGauss_netcdf_parallel_intel
Verifying: Test control_c48_intel
Verifying: Test control_c192_intel
Verifying: Test control_c384_intel
Verifying: Test control_c384gdas_intel
Verifying: Test control_stochy_intel
Verifying: Test control_stochy_restart_intel
Verifying: Test control_lndp_intel
Verifying: Test control_iovr4_intel
Verifying: Test control_iovr5_intel
Verifying: Test control_p8_intel
Verifying: Test control_p8_ugwpv1_intel
Verifying: Test control_restart_p8_intel
Verifying: Test control_noqr_p8_intel
Verifying: Test control_restart_noqr_p8_intel
Verifying: Test control_decomp_p8_intel
Verifying: Test control_p8_lndp_intel
Verifying: Test control_p8_rrtmgp_intel
Verifying: Test control_p8_mynn_intel
Verifying: Test merra2_thompson_intel
Verifying: Test regional_control_intel
Verifying: Test regional_restart_intel
Verifying: Test regional_decomp_intel
Verifying: Test regional_noquilt_intel
Verifying: Test regional_netcdf_parallel_intel
Verifying: Test regional_2dwrtdecomp_intel
Verifying: Test regional_wofs_intel
Verifying: Compile rrfs_intel
Verifying: Test rap_control_intel
Verifying: Test regional_spp_sppt_shum_skeb_intel
Verifying: Test rap_decomp_intel
Verifying: Test rap_restart_intel
Verifying: Test rap_sfcdiff_intel
Verifying: Test rap_sfcdiff_decomp_intel
Verifying: Test rap_sfcdiff_restart_intel
Verifying: Test hrrr_control_intel
Verifying: Test hrrr_control_decomp_intel
Verifying: Test hrrr_control_2threads_intel
Verifying: Test hrrr_control_restart_intel
Verifying: Test rrfs_v1beta_intel
Verifying: Test rrfs_v1nssl_intel
Verifying: Test rrfs_v1nssl_nohailnoccn_intel
Verifying: Compile csawmg_intel
Verifying: Test control_csawmg_intel
Verifying: Test control_csawmgt_intel
Verifying: Test control_ras_intel
Verifying: Compile wam_intel
Verifying: Test control_wam_intel
Verifying: Compile atm_faster_dyn32_intel
Verifying: Test control_p8_faster_intel
Verifying: Test regional_control_faster_intel
Verifying: Compile atm_debug_dyn32_intel
Verifying: Test control_CubedSphereGrid_debug_intel
Verifying: Test control_wrtGauss_netcdf_parallel_debug_intel
Verifying: Test control_stochy_debug_intel
Verifying: Test control_lndp_debug_intel
Verifying: Test control_csawmg_debug_intel
Verifying: Test control_csawmgt_debug_intel
Verifying: Test control_ras_debug_intel
Verifying: Test control_diag_debug_intel
Verifying: Test control_debug_p8_intel
Verifying: Test regional_debug_intel
Verifying: Test rap_control_debug_intel
Verifying: Test hrrr_control_debug_intel
Verifying: Test hrrr_gf_debug_intel
Verifying: Test hrrr_c3_debug_intel
Verifying: Test rap_unified_drag_suite_debug_intel
Verifying: Test rap_diag_debug_intel
Verifying: Test rap_cires_ugwp_debug_intel
Verifying: Test rap_unified_ugwp_debug_intel
Verifying: Test rap_lndp_debug_intel
Verifying: Test rap_progcld_thompson_debug_intel
Verifying: Test rap_noah_debug_intel
Verifying: Test rap_sfcdiff_debug_intel
Verifying: Test rap_noah_sfcdiff_cires_ugwp_debug_intel
Verifying: Test rrfs_v1beta_debug_intel
Verifying: Test rap_clm_lake_debug_intel
Verifying: Test rap_flake_debug_intel
Verifying: Test gnv1_c96_no_nest_debug_intel
Verifying: Compile wam_debug_intel
Verifying: Test control_wam_debug_intel
Verifying: Compile rrfs_dyn32_phy32_intel
Verifying: Test regional_spp_sppt_shum_skeb_dyn32_phy32_intel
Verifying: Test rap_control_dyn32_phy32_intel
Verifying: Test hrrr_control_dyn32_phy32_intel
Verifying: Test hrrr_control_decomp_dyn32_phy32_intel
Verifying: Test rap_restart_dyn32_phy32_intel
Verifying: Test hrrr_control_restart_dyn32_phy32_intel
Verifying: Compile rrfs_dyn32_phy32_faster_intel
Verifying: Test conus13km_control_intel
Verifying: Test conus13km_2threads_intel
Verifying: Test conus13km_restart_mismatch_intel
Verifying: Compile rrfs_dyn64_phy32_intel
Verifying: Test rap_control_dyn64_phy32_intel
Verifying: Compile rrfs_dyn32_phy32_debug_intel
Verifying: Test rap_control_debug_dyn32_phy32_intel
Verifying: Test hrrr_control_debug_dyn32_phy32_intel
Verifying: Test conus13km_debug_intel
Verifying: Test conus13km_radar_tten_debug_intel
Verifying: Compile rrfs_dyn64_phy32_debug_intel
Verifying: Test rap_control_dyn64_phy32_debug_intel
Verifying: Compile hafsw_intel
Verifying: Test hafs_regional_atm_intel
Verifying: Test hafs_regional_atm_thompson_gfdlsf_intel
Verifying: Test hafs_regional_atm_ocn_intel
Verifying: Test hafs_regional_atm_wav_intel
Verifying: Test hafs_regional_atm_ocn_wav_intel
Verifying: Test hafs_regional_1nest_atm_intel
Verifying: Test hafs_regional_telescopic_2nests_atm_intel
Verifying: Test hafs_global_1nest_atm_intel
Verifying: Test hafs_global_multiple_4nests_atm_intel
Verifying: Test hafs_regional_specified_moving_1nest_atm_intel
Verifying: Test hafs_regional_storm_following_1nest_atm_intel
Verifying: Test hafs_regional_storm_following_1nest_atm_ocn_intel
Verifying: Test hafs_global_storm_following_1nest_atm_intel
Verifying: Test gnv1_nested_intel
Verifying: Compile hafsw_debug_intel
Verifying: Test hafs_regional_storm_following_1nest_atm_ocn_debug_intel
Verifying: Compile hafsw_faster_intel
Verifying: Test hafs_regional_storm_following_1nest_atm_ocn_wav_intel
Verifying: Compile hafs_all_intel
Verifying: Test hafs_regional_docn_intel
Verifying: Test hafs_regional_docn_oisst_intel
Verifying: Test hafs_regional_datm_cdeps_intel
Verifying: Compile datm_cdeps_intel
Verifying: Test datm_cdeps_control_cfsr_intel
Verifying: Test datm_cdeps_restart_cfsr_intel
Verifying: Test datm_cdeps_control_gefs_intel
Verifying: Test datm_cdeps_iau_gefs_intel
Verifying: Test datm_cdeps_stochy_gefs_intel
Verifying: Test datm_cdeps_ciceC_cfsr_intel
Verifying: Test datm_cdeps_bulk_cfsr_intel
Verifying: Test datm_cdeps_bulk_gefs_intel
Verifying: Test datm_cdeps_mx025_cfsr_intel
Verifying: Test datm_cdeps_mx025_gefs_intel
Verifying: Test datm_cdeps_multiple_files_cfsr_intel
Verifying: Test datm_cdeps_3072x1536_cfsr_intel
Verifying: Test datm_cdeps_gfs_intel
Verifying: Compile datm_cdeps_debug_intel
Verifying: Test datm_cdeps_debug_cfsr_intel
Verifying: Compile datm_cdeps_faster_intel
Verifying: Test datm_cdeps_control_cfsr_faster_intel
Verifying: Compile datm_cdeps_land_intel
Verifying: Test datm_cdeps_lnd_gswp3_intel
Verifying: Test datm_cdeps_lnd_gswp3_rst_intel
Verifying: Compile atml_intel
Verifying: Test control_p8_atmlnd_sbs_intel
Verifying: Compile atmw_intel
Verifying: Test atmwav_control_noaero_p8_intel
Verifying: Compile atmwm_intel
Verifying: Test control_atmwav_intel
Verifying: Compile atmaero_intel
Verifying: Test atmaero_control_p8_intel
Verifying: Test atmaero_control_p8_rad_intel
Verifying: Test atmaero_control_p8_rad_micro_intel
Verifying: Compile atmaq_intel
Verifying: Test regional_atmaq_intel
Verifying: Compile atmaq_faster_intel
Verifying: Test regional_atmaq_faster_intel

REGRESSION TEST WAS SUCCESSFUL
```
The above is output from the new (current state of) RegressionTests_pretest.log with a successful test.

**Comment by DeniseWorthen** [comment]
Don't we just need to say that something failed, rather than listing everything that passed? So if everything is OK

```
All compilations were successful
All tests ran to completion
All baseline comparisons were successful
```

And only if something fails, logging specifics about what did not work (a compile, a run, a comparison).

**Comment by BrianCurtis-NOAA** [comment]
Part of the verification function is to spit out more if there's a failure/missing test/compile. I think having the output above is important, and it cuts down a lot from the verbose RegressionTests_\<machine\>.log

**Comment by DeniseWorthen** [comment]
Hm, well maybe others can chime in. I find the above too verbose. From the perspective of someone making a PR, all I care about is what doesn't work---either it fails to compile, it fails to run or it fails to compare. No news is good news.

**Comment by junwang-noaa** [comment]
Maybe just me, I feel it's convenient to see a list of verified tests to have a full picture.

**Comment by BrianCurtis-NOAA** [comment]
I'm making some adjustments and i'll get an example with failed tests soon.

**Comment by DusanJovic-NOAA** [comment]
What's the fundamental difference between this log file format and the current one, other than the file size? This file shows practically the same information except the list of each output file for each test (which in my opinion is useful information). Are we trying to reduce the size of the log file?

**Comment by DeniseWorthen** [comment]
For me, there are a couple of goals. This would be used **only** for pre-test confirmation that the developer a) ran the full RT and b) has the expected result (either tests changed or didn't change). It would automatically commit this to their branch. 

I'm not suggesting we don't see everything when we go to actual commit time. But for this purpose, I think a summary is good enough. I think having a long list of tests and their status is a problem---it is too easy to overlook a test which actually didn't run (see Issue 1821).

**Comment by BrianCurtis-NOAA** [comment]
This will not commit to their branch automatically. I can't write that code into it, as NCO will block it on WCOSS2.

**Comment by DusanJovic-NOAA** [comment]
Current log file lists the failed tests at the end of the log and it prints the message that the test is not successful. That issues in 1821 is something that needs to be fixed (and will be fixed) but that's not the problem of a log file format. Having two different log formats will just complicate the scripts for very little benefit, unless the only reason is to save on the file size, which is not a lot if we still end up committing the logs from all supported platforms in the current format anyway, so we are talking about only one file.

**Comment by DeniseWorthen** [comment]
OK, so I'm in a minority. If people want the whole list of files and their comparisons that is fine. But I think we do need an explicit summary that all compilations succeeded (or did not), all tests ran to completion (or did not) and all tests were b4b (or were not).

**Comment by BrianCurtis-NOAA** [comment]
For non-pretest logs, it will have the same information, but in a different order. The order will match the rt.conf (or what's defined in $TESTS_FILE). The pre-test will be smaller and only show information for failed tests, but still have information on all tests/compiles.

I'm open to adjusting things, but set that we do need a defined method for pre-test so we can just always say "run ./rt.sh -p" or ask have you run "rt.sh -p" yet. Let me get a PR draft up so you can see the changes and also comment on them.

**Comment by DeniseWorthen** [comment]
> This will not commit to their branch automatically. I can't write that code into it, as NCO will block it on WCOSS2.

We don't ask for pre-commit logs from WCOSS2 (no gnu there). But are you saying the code itself can't be in the script?

**Comment by DusanJovic-NOAA** [comment]
I guess I'm confused about this concept of 'pre-test'. Isn't the:

`./rt.sh -e`

executed on Hera effectively a 'pre-test'.

The log file created by the test should show either successful test message in which case ALL tests are b4b (nccmp b4b) with the current baselines or test failed in which case at the end of the log file there should be a list of failed test. When we find that this is not the case we should simply fix the script (as in 1821)

**Comment by BrianCurtis-NOAA** [comment]
> > This will not commit to their branch automatically. I can't write that code into it, as NCO will block it on WCOSS2.
> 
> We don't ask for pre-commit logs from WCOSS2 (no gnu there). But are you saying the code itself can't be in the script?

From what I understand, thats correct.

**Comment by BrianCurtis-NOAA** [comment]
@DusanJovic-NOAA The idea behind the pre-test is so we can point developers to one command to give us the information we need for each PR before we allow it into the commit queue.

I've re-written things to give us a central location for all information coming out of the testing process, so we can reduce the log information for pre-test and now if needed we can add new information or limit it even more.

**Comment by DeniseWorthen** [comment]
@DusanJovic-NOAA I think originally the idea was that this could get committed automatically to a user's branch, but it sounds like that is not possible. So, yes, it is effectively now back to just asking developers to run/post. So we're back really to just fixing 1821.

**Comment by BrianCurtis-NOAA** [comment]
> @DusanJovic-NOAA I think originally the idea was that this could get committed automatically to a user's branch, but it sounds like that is not possible. So, yes, it is effectively now back to just asking developers to run/post. So we're back really to just fixing 1821.

I disagree. This log "RegressionTests_pretest.log" will provide the repo with a history of proof of how PR's are changing tests. Yes they still have to post it, "git add RegressionTests_pretest.log && git commit -m 'pretest log' && push origin <branch>"

**Comment by DusanJovic-NOAA** [comment]
> @DusanJovic-NOAA The idea behind the pre-test is so we can point developers to one command to give us the information we need for each PR before we allow it into the commit queue.
> 
> I've re-written things to give us a central location for all information coming out of the testing process, so we can reduce the log information for pre-test and now if needed we can add new information or limit it even more.

Isn't 
```
./rt.sh -e
```
that one command? I struggle to understand what 'pre-test' is doing differently other than creating different log file.

**Comment by DusanJovic-NOAA** [comment]
> > @DusanJovic-NOAA I think originally the idea was that this could get committed automatically to a user's branch, but it sounds like that is not possible. So, yes, it is effectively now back to just asking developers to run/post. So we're back really to just fixing 1821.
> 
> I disagree. This log "RegressionTests_pretest.log" will provide the repo with a history of proof of how PR's are changing tests. Yes they still have to post it, "git add RegressionTests_pretest.log && git commit -m 'pretest log' && push origin "

So now the question is whether the log file is posted to PR as a comment or committed to repository. If we are talking about 'proof' of which test are changed by a PR then we should ask PR authors to commit 'new_baseline' file which will allow code managers to use that file to create new baselines and verify that only those tests specified in the 'new_baseline' file are indeed the only tests that are modified. Which is much more useful information than a log file with a failed test list.

**Comment by DeniseWorthen** [comment]
Sorry, I said "run/post"....meant to say "run/commit". I think we've all learned that "posting" the log (ie attaching it to the template) just isn't working for the majority of developers.

And my concern is less about test history (that is covered by the actual log at time of commit). I'm mostly concerned about notifying developers that there is a problem w/ their PR. An unexpected test failed, it didn't compile, it failed to run (maybe a input file is missing, etc).

**Comment by BrianCurtis-NOAA** [comment]
@DusanJovic-NOAA I think I should add an output of the -p options to create a file to use with -b. I wonder if we could also add this file to the repository. Something like test_changes.out which would provide the repository with a history of such information.

**Comment by DusanJovic-NOAA** [comment]
Sorry, I still do not understand what is the motivation to add a new pre-test option (-p) and how it is different from let's say `./rt.sh -e`. Is the pre-test going to run different tests than regular (non pre-test) tests? Or is the only difference the content of the log file (shorter log file)?  

Regarding the file that will list the tests that are expected to change the baselines, I think that file should be created by developer(s) 'manually', not by simply running some scripts. If it's created by some script, how do we 'know' that it will list just the tests that developers expect to be changed and not other tests by some unintentional change?

**Comment by BrianCurtis-NOAA** [comment]
We've actually scrapped the -p idea for now. Check out the PR #2061 for an update. I'll update this issue accordingly.

In my opinion it's on the developers to confirm that the changes they see in the log match what they expect to change based on the changes in the PR. Once this is accepted by the developer, then the generated test_changes.out (or something else?) will have a proper list. This new file will be the repository's history for that.

**Comment by DusanJovic-NOAA** [comment]
Okay, so if this issue is not about pre-test (-p), then it's about the format and the content of the log file, is that correct?

I took a look at #2061, and in addition to simple renaming of some variables and removal of the lsf scheduler I see huge changes in rt.sh. The new function (generate_log) is almost 300 lines of code, that replaces ~35 lines of code that concatenates log results of individual tests and if there are failed tests it concatenates fail_test_* fail_compile_*.

I wonder if it's worth adding so much complexity. If the idea is to create a shorter log file, without listing each file comparison status, then the changes should be made in rt_utils.sh (in check_results function specifically), and in rt.sh we should only combine all individual test's logs, without adding any result checking logic. The rt.sh should be just a simple driver that submits tests, it should not have any code that parses individual test's log files.

The other issue I have with this approach is that this new  function (generate_log) parses rt.conf, which has already been parsed, which leads to a lot of duplicated code and logic, which is difficult to maintain. Almost half of generate_log is parsing logic of rt.conf. It also contains code cleanup code, not only log generation code which is also confusing.

**Comment by BrianCurtis-NOAA** [comment]
@DusanJovic-NOAA I think parsing rt.conf is beneficial in this case. It's the official information sent to rt.sh on what tests should be run. So in my opinion, to verify all tests, that makes sense.

I don't think generate_logs() is related to check_results() or should be any part of it. check_results() is tied to each individual test (it's called in run_test.sh), but bares no responsibility to the rt.sh as a whole. generate_logs() is more of a verification tool for the whole rt.sh that also takes on the responsibility of generating the logs and why I chose to put it in rt.sh.

I wouldn't say this is too complex, either. The extra lines of code do not slow down rt.sh, or impact any of the timings reported to the timings logs. rt.sh is ~1100 lines here, and ~990 in develop branch. We're really only adding ~100 lines. Now all log generation is contained in one function, makes it easier to find when changes are needed.

Now for the code-cleanup, that's what was in rt.sh already and I moved it into the generate_logs() function, we can separate that out or re-work the logic if you want, but that doesn't need to happen here. I'm happy to create an issue for that work if you prefer to separate it.
